kernel/src/error.rs

```diff
@@ -119,6 +133,43 @@ impl Error {
     pub fn join_failure(msg: impl ToString) -> Self {
         Self::JoinFailure(msg.to_string())
     }
+
+    // Capture a backtrace when the error is constructed.
+    #[must_use]
+    pub fn with_backtrace(self) -> Self {
+        let backtrace = Backtrace::capture();
+        match backtrace.status() {
+            BacktraceStatus::Captured => Self::Backtraced {
+                source: Box::new(self),
+                backtrace: Box::new(backtrace),
+            },
+            _ => self,
+        }
+    }
+}
+
+macro_rules! from_with_backtrace(
```

Comment:
Isn't this more code with the macro given you are only implementing for 2 error types?

Reply:
yes. the idea is that it's really easy in the future to add more if you need it to debug something